Cython/Utility/ModuleSetupCode.c

```diff
@@ -2231,6 +2231,105 @@ static PyObject* __Pyx_PyCode_New(
     return code_obj;
 }
 
+/////////////////////////// PyVersionSanityCheck.proto ///////////////////////
+
+static int __Pyx_VersionSanityCheck(void); /* proto */
+
+/////////////////////////// PyVersionSanityCheck ///////////////////////
+
+static int __Pyx_VersionSanityCheck(void) {
+  // Implementation notes:
+  //  The main thing I'm worried about in mixing up Py_GIL_DISABLED since this is incredibly
+  //  easy to do on Windows (where only a single pyconfig.h is provided, which is shared between
+  //  the freethreading and non-freethreading executables).  Getting this wrong instantly crashes.
+  //
+  //  The debug and trace-refs checks are just because they're easy to do at the same time.
+  //
+  //  In order to test it we cannot use any internals of Python objects from C
+  //  (especially refcounting) since we potentially have completely the wrong structure
+  //  of PyObject.  PyRun_SimpleStringFlags and PySys_GetObject both look fairly safe to
+  //  use like this.
+  //
+  //  Since PyRun_SimpleStringFlags runs in the __main__ scope, we should be careful not
+  //  to define variables inside it.
+  //  We use PyRun_SimpleStringFlags instead of PyRun_SimpleString since SimpleString is
+  //  a macro, and MSVC doesn't like #ifdef within a macro expansion.
+  //
+  //  PyRun_SimpleStringFlags prints and clears the exception information.  We raise
+  //  therefore raise the proper exception from C (I think the interface to PyErr_SetString
+  //  should be consistent enough between builds for this to be OK). This is fragile but
+  //  the best we can do.
+  //
+  #if CYTHON_COMPILING_IN_CPYTHON
+    if (PySys_GetObject("gettotalrefcount")) {
+      #ifndef Py_DEBUG
+        PyErr_SetString(
+            PyExc_ImportError,
+            "Module was compiled with a non-debug version of Python but imported into a debug version."
+        );
+        return -1;
+      #endif
+    } else {
+      #ifdef Py_DEBUG
+        PyErr_SetString(
+            PyExc_ImportError,
+            "Module was compiled with a debug version of Python but imported into a non-debug version."
+        );
+        return -1;
+      #endif
+    }
+    #if PY_VERSION_HEX >= 0x030d0000
+    if (PyRun_SimpleStringFlags(
+      "if "
+      #ifdef Py_GIL_DISABLED
+        "not "
+      #endif
+      "__import__('sysconfig').get_config_var('Py_GIL_DISABLED'):\n"
+      "  raise ImportError",
+      NULL
+    ) == -1) {
+        PyErr_SetString(
+            PyExc_ImportError,
+      #ifdef Py_GIL_DISABLED
+            "Module was compiled with a freethreading build of Python but imported into a non-freethreading build."
+      #else
+            "Module was compiled with a non-freethreading build of Python but imported into a freethreading build."
+      #endif
+        );
```

Comment:
You could make the exception message a part of the Python code (i.e. `raise ImportError('message')` to avoid raising the exception twice.

Reply:
That was my initial thought but unfortunately it doesn't work. `PyRun_SimpleStringFlags` clears the exception and there's no way to get it back. The interfaces that don't swallow the exception mostly require creating Python objects, which I'm trying to avoid if we're linked to the wrong ABI.